sdks/python/apache_beam/transforms/trigger.py

```diff
@@ -1038,14 +1094,21 @@ def process_timer(self, window_id, unused_name, time_domain, timestamp,
     window = state.get_window(window_id)
     if state.get_state(window, self.TOMBSTONE):
       return
-    if time_domain == TimeDomain.WATERMARK:
-      if not self.is_merging or window in state.known_windows():
-        context = state.at(window)
+
+    if not self.is_merging or window in state.known_windows():
+      context = state.at(window, self.clock)
+      if time_domain == TimeDomain.WATERMARK:
         if self.trigger_fn.should_fire(timestamp, window, context):
           finished = self.trigger_fn.on_fire(timestamp, window, context)
           yield self._output(window, finished, state)
-    else:
-      raise Exception('Unexpected time domain: %s' % time_domain)
+      elif time_domain == TimeDomain.REAL_TIME:
```

Comment:
<!--thread_id:cc_162184928_t; commit:a769afcabd2a4b8ad94cae1f1c62e73823e42d7b; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> We should disallow the setting of real time timers if `self.clock` is None.

<!--section:body-->
Good idea. Better than that, I am making the clock always have a value, instead of taking the default None.